File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3090_maximum-length-substring-with-two-occurrences/solutions/python3/solution.py
class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        left = 0
        max_len = 0
        char_count = {}

        for right in range(len(s)):
            char_count[s[right]] = char_count.get(s[right], 0) + 1
            
            # Shrink the window until each character has at most two occurrences
            while any(c > 2 for c in char_count.values()):
                char_count[s[left]] -= 1
                if char_count[s[left]] == 0:
                    del char_count[s[left]]
                left += 1
            
            max_len = max(max_len, right - left + 1)

        return max_len

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3090_maximum-length-substring-with-two-occurrences/solutions/java/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maximumLengthSubstring(String s) {
        Map<Character, Integer> countMap = new HashMap<>();
        int maxLen = 0;
        int start = 0;

        for (int end = 0; end < s.length(); end++) {
            char currentChar = s.charAt(end);
            countMap.put(currentChar, countMap.getOrDefault(currentChar, 0) + 1);

            while (countMap.get(currentChar) > 2) {
                char startChar = s.charAt(start);
                countMap.put(startChar, countMap.get(startChar) - 1);
                start++;
            }

            maxLen = Math.max(maxLen, end - start + 1);
        }
        
        return maxLen;
    }
}
